File: scripts/check-translations.ts
#!/usr/bin/env node
/**
 * Translation Check - Main Entry Point
 *
 * Static analysis tool that detects missing translations in the R-Instat
 * VB.NET codebase. Read-only: reports drift and exits non-zero when any
 * strings in code are missing from the EN JSON files.
 *
 * Usage:
 *   npm run check-translations              # Check all VB.NET files, exit 1 if any missing
 *   npm run check-translations:ci           # Check only changed files, GitHub annotations
 *   npm run check-translations:verbose      # Verbose output
 *
 * Options:
 *   --ci          CI mode: only check changed files, use GitHub-friendly output
 *   --verbose     Show detailed information including pattern matches
 *   --json        Output as JSON
 *   --help        Show help
 *
 * To resolve missing translations locally: npm run sync-translations
 */

import * as path from 'path';

import { getProjectRoot } from './lib/project-root.js';
import { isCI } from './lib/git-utils.js';
import { runScan } from './lib/scan.js';
import { findOrphans } from './lib/orphan-finder.js';
import {
  createSummary,
  generateReport,
  writeReportToFile,
  generateConsoleSummary,
  ReportOptions
} from './lib/reporter.js';

interface CliOptions {
  ciMode: boolean;
  verbose: boolean;
  jsonOutput: boolean;
  help: boolean;
  files: string[];
}

function parseArgs(args: string[]): CliOptions {
  const options: CliOptions = {
    ciMode: false,
    verbose: false,
    jsonOutput: false,
    help: false,
    files: []
  };

  for (const arg of args) {
    switch (arg) {
      case '--ci':
        options.ciMode = true;
        break;
      case '--verbose':
      case '-v':
        options.verbose = true;
        break;
      case '--json':
        options.jsonOutput = true;
        break;
      case '--help':
      case '-h':
        options.help = true;
        break;
      default:
        if (!arg.startsWith('-')) {
          options.files.push(arg);
        }
        break;
    }
  }

  if (isCI()) {
    options.ciMode = true;
  }

  return options;
}

function printHelp(): void {
  console.log(`
Translation Check - Static analysis for missing translations

Usage:
  npx ts-node check-translations.ts [options] [files...]

Options:
  --ci          CI mode: only check changed files, use GitHub-friendly output
  --verbose     Show detailed information including pattern matches
  --json        Output as JSON
  --help        Show this help message

Exits non-zero if any strings are missing from the EN JSON files.
Run \`npm run sync-translations\` locally to append missing strings.
`);
}

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  const options = parseArgs(args);

  if (options.help) {
    printHelp();
    process.exit(0);
  }

  let baseDir: string;
  try {
    baseDir = await getProjectRoot();
  } catch (error) {
    console.error((error as Error).message);
    process.exit(1);
  }

  if (options.verbose) {
    console.log(`Base directory: ${baseDir}`);
    console.log(`CI mode: ${options.ciMode}`);
  }

  const { violations, stats, extractedStrings, translations } = await runScan(baseDir, {
    ciMode: options.ciMode,
    verbose: options.verbose,
    files: options.files
  });

  const orphanAnalysis = findOrphans(extractedStrings, translations);
  const totalTranslationKeys = translations.allKeys.size;

  const summary = createSummary(
    violations,
    stats,
    options.ciMode,
    orphanAnalysis,
    totalTranslationKeys
  );

  const reportPath = path.join(baseDir, 'scripts', 'translation-report.json');
  writeReportToFile(summary, reportPath);

  if (options.jsonOutput) {
    console.log(JSON.stringify(summary, null, 2));
  } else if (options.ciMode) {
    const reportOptions: ReportOptions = { format: 'github', verbose: options.verbose };
    console.log(generateReport(summary, reportOptions));
  } else {
    console.log(generateConsoleSummary(summary, reportPath));
    if (violations.length > 0) {
      console.log('');
      console.log('To append missing strings to the EN JSON files, run:');
      console.log('  npm run sync-translations');
    }
  }

  // Unified behavior: fail on any missing translation, local or CI.
  // Dev workflow: run `npm run sync-translations` to resolve, then re-run.
  if (violations.length > 0) {
    process.exit(1);
  }

  process.exit(0);
}

main().catch(error => {
  console.error('Error:', error);
  process.exit(1);
});

File: scripts/check_translations.py
#!/usr/bin/env python3
"""Translation check entry point.

Read-only static analysis that fails when any strings in VB.NET code are
missing from EN translation JSON files.
"""

from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

from translation_common import (
    REPORT_END_MARKER,
    REPORT_START_MARKER,
    create_summary,
    find_orphans,
    find_project_root,
    find_uncovered_text_setters,
    generate_console_summary,
    generate_github_report,
    is_ci,
    run_scan,
    write_report_to_file,
)


def parse_args(argv: list[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Translation Check - Static analysis for missing translations"
    )
    parser.add_argument(
        "--ci",
        action="store_true",
        help="CI mode: check changed files and emit GitHub-friendly output",
    )
    parser.add_argument(
        "--verbose",
        "-v",
        action="store_true",
        help="Show detailed information",
    )
    parser.add_argument("--json", action="store_true", help="Output summary as JSON")
    parser.add_argument(
        "--base",
        default=None,
        help="Base ref for --ci diff (default: $GITHUB_BASE_REF or master). "
             "Use e.g. --base=origin/master locally to reproduce CI.",
    )
    parser.add_argument("files", nargs="*", help="Optional explicit VB files to scan")
    return parser.parse_args(argv)


def main(argv: list[str]) -> int:
    options = parse_args(argv)
    ci_mode = options.ci or is_ci()

    try:
        base_dir = find_project_root()
    except RuntimeError as exc:
        print(str(exc), file=sys.stderr)
        return 1

    if options.verbose:
        print(f"Base directory: {base_dir}")
        print(f"CI mode: {ci_mode}")

    try:
        scan_result = run_scan(
            base_dir,
            ci_mode=ci_mode,
            verbose=options.verbose,
            files=options.files,
            base_branch=options.base,
        )
    except RuntimeError as exc:
        if ci_mode:
            print(REPORT_START_MARKER)
            print("## Translation Check Results")
            print("")
            print("❌ **Translation check failed before scan completed.**")
            print("")
            print(f"Error: `{str(exc)}`")
            print(REPORT_END_MARKER)
        else:
            print(str(exc), file=sys.stderr)
        return 1

    violations = scan_result["violations"]
    stats = scan_result["stats"]
    extracted_strings = scan_result["extractedStrings"]
    translations = scan_result["translations"]

    orphan_analysis = find_orphans(extracted_strings, translations)
    total_translation_keys = len(translations.all_keys)

    summary = create_summary(
        violations,
        stats,
        ci_mode,
        orphan_analysis,
        total_translation_keys,
    )

    report_path = base_dir / "scripts" / "translation-report.json"
    write_report_to_file(summary, report_path)

    if options.json:
        print(json.dumps(summary, indent=2))
    elif ci_mode:
        print(generate_github_report(summary))
    else:
        print(generate_console_summary(summary, report_path))
        if summary["totalMissingTranslations"] > 0:
            print("")
            print("To append missing strings to the EN JSON files, run:")
            print("  python scripts/sync_translations.py")

    uncovered = find_uncovered_text_setters(base_dir)
    if uncovered:
        print("", file=sys.stderr)
        print("error: extractor coverage gap detected.", file=sys.stderr)
        print(
            "The following Set*Text(...) methods are used in VB code but have no "
            "extractor regex in translation_common.py. Strings passed to them are "
            "silently invisible to check/sync. Add a regex and update KNOWN_TEXT_SETTERS:",
            file=sys.stderr,
        )
        for name, count in uncovered:
            print(f"  - {name} ({count} call sites)", file=sys.stderr)
        return 1

    if summary["totalMissingTranslations"] > 0:
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))
